ir/memory.cpp

```diff
@@ -296,79 +416,70 @@ void Memory::free(const expr &ptr) {
   blocks_liveness = blocks_liveness.store(p.get_bid(), false);
 }
 
-void Memory::store(const expr &p, const StateValue &v, Type &type,
+void Memory::store(const expr &p, const StateValue &v, const Type &type,
                    unsigned align) {
-  StateValue val = type.toBV(v);
-  unsigned bits = val.bits();
-  unsigned bytes = divide_up(bits, 8);
-
-  val = val.zext(bytes * 8 - bits);
-
   Pointer ptr(*this, p);
-  ptr.is_dereferenceable(bytes, align);
 
-  for (unsigned i = 0; i < bytes; ++i) {
-    // FIXME: right now we store in little-endian; consider others?
-    expr data  = val.value.extract((i + 1) * 8 - 1, i * 8);
-    expr np    = val.non_poison.extract((i + 1) * 8 - 1, i * 8);
-    blocks_val = blocks_val.store((ptr + i).release(), np.concat(data));
+  vector<Byte> bytes;
+  valueToBytes(v, type, *this, bytes);
+  for (unsigned i = 0; i < bytes.size(); ++i) {
+    // FIXME: support big-endian
+    blocks_val = blocks_val.store((ptr + i).release(), bytes[i]());
   }
+  ptr.is_dereferenceable(bytes.size(), align);
 }
 
-StateValue Memory::load(const expr &p, Type &type, unsigned align) {
-  auto bits = type.bits();
-  unsigned bytes = divide_up(bits, 8);
+StateValue Memory::load(const expr &p, const Type &type, unsigned align) {
   Pointer ptr(*this, p);
-  ptr.is_dereferenceable(bytes, align);
 
-  StateValue val;
-  bool first = true;
+  bool isPtrTy = type.isPtrType();
+  auto bitsize = isPtrTy ? bitsMemorySize() : type.bits();
```

Comment:
I'm confused why this is needed. Is PtrType.bits() wrong?

Reply:
`PtrType.bits()` returns 64 + 8 + 8, so bitsPtrSize() which returns 64 was needed.